# Ejercicio 2 analisis de datos

In [1]:
# Importo librerias
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

In [2]:
import pandas as pd

# URL del archivo en tu repositorio de GitHub
url = "https://raw.githubusercontent.com/Jevillalbav/parcial_final_poo/main/airlines_reviews.csv"

# Leer el archivo directamente usando pandas
df = pd.read_csv(url)
df.columns = [i.replace(' ', '_').lower() for i in df.columns]

-------------------

Se sugiere que antes de comenzar con el análisis se lea cada una de las descripciones de
las columnas para tener un contexto general de los datos.

Primero, se requiere poder dar respuesta a lo siguiente:

    a. ¿Cuántos comentarios se encuentran verificados?
    b. ¿Cuál es el rango de fechas de los vuelos que se registran en el dataset?
    c. ¿Cuáles aerolíneas registran por lo menos un comentario?
    d. ¿Cuál es el promedio de comentarios de cada aerolínea por año?


In [3]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8100 entries, 0 to 8099
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   title                   8100 non-null   object
 1   name                    8100 non-null   object
 2   review_date             8100 non-null   object
 3   airline                 8100 non-null   object
 4   verified                8100 non-null   object
 5   reviews                 8100 non-null   object
 6   type_of_traveller       8100 non-null   object
 7   month_flown             8100 non-null   object
 8   route                   8100 non-null   object
 9   class                   8100 non-null   object
 10  seat_comfort            8100 non-null   int64 
 11  staff_service           8100 non-null   int64 
 12  food_&_beverages        8100 non-null   int64 
 13  inflight_entertainment  8100 non-null   int64 
 14  value_for_money         8100 non-null   int64 
 15  over

In [4]:
df.isna().sum()

title                     0
name                      0
review_date               0
airline                   0
verified                  0
reviews                   0
type_of_traveller         0
month_flown               0
route                     0
class                     0
seat_comfort              0
staff_service             0
food_&_beverages          0
inflight_entertainment    0
value_for_money           0
overall_rating            0
recommended               0
dtype: int64

In [5]:
df.sample(7).T

2451  \
title                                    very good business class product   
name                                                              B Virgo   
review_date                                                    2017-01-08   
airline                                                     Qatar Airways   
verified                                                            False   
reviews                  Melbourne to Casablanca via Doha, returning f...   
type_of_traveller                                            Solo Leisure   
month_flown                                                  January 2017   
route                                                  BCN to ADL via DOH   
class                                                      Business Class   
seat_comfort                                                            5   
staff_service                                                           5   
food_&_beverages                                                        3   
inflight_entertainment                                                  5   
value_for_money                                                         4   
overall_rating                                                          6   
recommended                                                           yes   

                                                                     2047  \
title                                   amazed by the attention to detail   
name                                                             S Norton   
review_date                                                    2018-09-03   
airline                                                     Qatar Airways   
verified                                                             True   
reviews                  Flew Doha to Goa. Great airline, with a great...   
type_of_traveller                                            Solo Leisure   
month_flown                                                September 2018   
route                                         Milan to Cape Town via Doha   
class                                                       Economy Class   
seat_comfort                                                            3   
staff_service                                                           5   
food_&_beverages                                                        5   
inflight_entertainment                                                  4   
value_for_money                                                         5   
overall_rating                                                         10   
recommended                                                           yes   

                                                                     1579  \
title                   My experience with Qatar Airways was just exce...   
name                                                             N Sakala   
review_date                                                    2021-02-17   
airline                                                     Qatar Airways   
verified                                                             True   
reviews                  My experience with Qatar Airways was just exc...   
type_of_traveller                                            Solo Leisure   
month_flown                                                 February 2021   
route                                      Seoul to Johannesburg via Doha   
class                                                       Economy Class   
seat_comfort                                                            5   
staff_service                                                           5   
food_&_beverages                                                        5   
inflight_entertainment                                                  5   
value_for_money                                                         5   
overall_rating                                                 

-----------

## ¿ Cuantos comentarios se encuentran verificados ? 

In [6]:
## se cuentan valores
df['verified'].value_counts()

True                                                                                                                                                                                                                                           6216
False                                                                                                                                                                                                                                          1881
*Unverified*                                                                                                                                                                                                                                      1
NotVerified                                                                                                                                                                                                                                       1
we do appreciate you bri

In [65]:
# se revisa que paso con los que no son True o False
df.query('verified != "True" & verified != "False"')


title          name review_date  \
2225     CSD was professional but friendly        D Webb  2017-11-18   
3358                     Very disappointed        HY Teo  2019-08-13   
3822  they allocated us to different seats  John Perrins  2017-11-19   

            airline                                           verified  \
2225  Qatar Airways                                       *Unverified*   
3358       Emirates                                        NotVerified   
3822       Emirates  we do appreciate you bringing this matter to o...   

                                                reviews type_of_traveller  \
2225    Delhi to London via Doha. The return flight ...      Solo Leisure   
3358    Singapore to Barcelona via Dubai with Emirat...    Couple Leisure   
3822    Birmingham to Johannesburg via Dubai. We pai...    Couple Leisure   

         month_flown                             route          class  \
2225   November 2017                      Kiev to Doha    First Class   
3358     August 2019  Melbourne to Frankfurt via Dubai  Economy Class   
3822  September 2017             Narita Tokyo to Dubai  Economy Class   

      seat_comfort  staff_service  food_&_beverages  inflight_entertainment  \
2225             5              4                 5                       4   
3358             3              3                 1                       2   
3822             2              5                 4                       4   

      value_for_money  overall_rating recommended  
2225                3               7         yes  
3358                3               5          no  
3822                1               2          no

In [7]:
conditions = [
    df['verified'] == 'True',
    df['verified'] == 'False',
    df['verified'] == 'NotVerified',
    df['verified'] == '*Unverified*']
    
df['verified_'] = np.select(conditions, ['True', 'False', 'False', 'False'], default= np.nan)

In [8]:
df['verified_'].value_counts().to_frame()

verified_
True        6216
False       1883
nan            1

In [9]:
df['verified_'].value_counts(normalize=True).to_frame().round(2)

verified_
True        0.77
False       0.23
nan         0.00

## Respuesta:  
 
Los 3 casos diferentes de True o False en la columna verified fueron revisados, dos de ellos son unverified y not verified por lo cual es facil asumir que estos no han sido verificados, el tercer caso es distinto por lo cual no se asume nada y se da como mismatched.  

verified_

**True**	6216 casos

**False**	1883 casos 

**nan**	1 casos

-----------

## Cual es el rango de fechas de los vuelos que se registran en el dataset?

In [10]:
dates = df['month_flown'].str.split(' ', expand=True).copy()
dates.columns = ['month', 'year']
dates['month'] = dates['month'].map({'January': 1, 'February': 2, 'March': 3, 'April': 4, 'May': 5, 'June': 6, 'July': 7, 'August': 8, 'September': 9, 'October': 10, 'November': 11, 'December': 12})
dates['year'] = dates['year'].astype(int)
dates['date'] = pd.to_datetime(dates.assign(day=1)[['year', 'month', 'day']])

first = dates['date'].min()
last = dates['date'].max()

print(f'First date: {first}')
print(f'Last date: {last}')

First date: 2013-03-01 00:00:00
Last date: 2024-03-01 00:00:00


## Respuesta:  

El rango de fechas esta entre Marzo de 2013 y Marzo de 2024

----------------

## ¿Cuáles aerolíneas registran por lo menos un comentario?

In [11]:
df.groupby('airline').size().sort_values(ascending=False).to_frame().rename(columns={0: 'review'})

review
airline                       
Turkish Airlines          1685
Qatar Airways             1624
Emirates                  1350
Singapore Airlines         972
Air France                 798
Cathay Pacific Airways     744
EVA Air                    281
All Nippon Airways         258
Japan Airlines             201
Korean Air                 187

----------------------

## ¿Cuál es el promedio de comentarios de cada aerolínea por año?

*** la pregunta esta mal planteada, pero se entiende que lo que se quiere buscar es el promedio del rating de cada aerolinea por año, ya que no se puede calcular el promedio de un comentario ya que no es algo numerico, en el mejor de los casos se puede hacer un **conteo** de reseñas o un promedio de calificacion de estas, para utilidad del negocio se procede hacer el promedio de calificacion de cada aerolinea por año. ***

In [12]:
x = df[['review_date', 'airline', 'overall_rating']].copy()
x['review_date'] = pd.to_datetime(x['review_date'])
x['year'] = x['review_date'].dt.year.astype(int)

x = x.groupby(['year', 'airline'])[['overall_rating']].mean().reset_index().pivot_table(index='airline', columns='year', values='overall_rating').sort_values(2024, ascending=False)
x

year                        2016      2017      2018      2019      2020  \
airline                                                                    
Korean Air              7.272727  6.733333  5.948718  6.457143  5.900000   
Japan Airlines          8.031250  7.142857  6.034483  7.705882  6.307692   
Singapore Airlines      7.254717  7.018293  7.143791  6.723810  5.608696   
Qatar Airways           7.411348  7.611321  7.164319  6.707113  8.170213   
Cathay Pacific Airways  7.153061  7.305882  6.715976  5.126506  3.744186   
All Nippon Airways      7.785714  8.500000  8.200000  7.340000  8.583333   
Air France              5.507937  5.629630  4.507042  4.593750  4.204082   
Emirates                6.171598  5.437736  4.328185  4.514286  3.922330   
Turkish Airlines        6.462185  5.595000  4.194915  3.589404  2.493333   
EVA Air                 7.892857  8.393939  8.112903  6.916667  7.136364   

year                        2021      2022      2023      2024  
airline                                                         
Korean Air              5.000000  5.111111  6.909091  8.666667  
Japan Airlines          7.272727  6.739130  6.857143  8.000000  
Singapore Airlines      4.357143  5.129630  6.096774  7.333333  
Qatar Airways           8.243902  6.682171  6.663551  5.966667  
Cathay Pacific Airways  1.000000  2.400000  5.702703  5.714286  
All Nippon Airways      8.666667  8.411765  7.717949  5.428571  
Air France              4.000000  4.176991  4.144068  3.875000  
Emirates                3.460000  4.024000  3.948718  3.352941  
Turkish Airlines        2.755208  2.782609  2.721461  2.675676  
EVA Air                 2.200000  4.944444  6.812500       NaN

## Respuesta:  

En este momento, la aerolinea mas destacada es **Korean Air** con un promedio de 8.6 sobre 10 para el año 2024. La peor calificada en lo que va del año es **Turkish Airlines** con un promedio de 2.67 sobre 10.

----------------

# Segunda Parte: 

    a. Un gráfico de barras que muestre la distribución de los vuelos a lo largo de los
        meses. Esto con el fin de poder identificar en cuales meses se viajan más o menos.

    b. Un histograma que muestre “Overall Rating” de las dos aerolíneas con mayor
    cantidad de comentarios. Se debe diferenciar por color estas dos compañías.
    Gracias a esto se de pronto se puede identificar si alguna de estas dos empresas
    tiene una mejor calificación respecto a la otra.

    c. Un gráfico tipo torta/pastel que muestre el porcentaje de comentarios de
    “Economy Class”, “Business Class” y “Others”, es decir, solo se deben poder
    observar en el gráfico 3 categorías.

    d. Un gráfico de dispersión que compare “Seat Comfort” y “Value for Money”.

---------------

## a. Un gráfico de barras que muestre la distribución de los vuelos a lo largo de los meses. Esto con el fin de poder identificar en cuales meses se viajan más o menos.

In [13]:
w = dates.groupby('month').size()

fig = px.bar(w, x=w.index, y=w.values, labels={'x': 'Month', 'y': 'Flights'}, title='Flights per month')
fig.update_layout(width=900, height=500, yaxis = dict( range=[500, 800]))
fig.show()

## Respuesta:  

Hay una clara tendencia al alza en los meses de Diciembre y Enero, esto tiene sentido con la teporada de navidad alrededor de estas fechas.A partir del mes de marzo se nota una disminucion en las cantidades de vueloos, alcanzando un minimo de vuelos en el mes de junio, el cual es un 26% menor al mes donde se realizarion mas viajes (Diciembre).

--------------

## b. Un histograma que muestre “Overall Rating” de las dos aerolíneas con mayor cantidad de comentarios. Se debe diferenciar por color estas dos compañías. Gracias a esto se de pronto se puede identificar si alguna de estas dos empresas tiene una mejor calificación respecto a la otra.

In [48]:
filtro = df['airline'].value_counts().head(2).index

c = df.query('airline in @filtro')[['airline', 'overall_rating']]

fig = px.histogram(c, x='overall_rating', color='airline', nbins= 10, barmode='overlay', title='Rating distribution by airline', color_discrete_map={'Qatar Airways': 'blue', 'Turkish Airlines': 'red'},
                   marginal='box', labels={'overall_rating': 'Rating', 'airline': 'Airline'})
fig.update_layout(width=900, height=500 )
fig.show()

## Respuesta:

Hay una clara diferencia entre las dos aerolineas mas reseñadas, **Qatar Airways** y **Turkish Airlines**, la primera tiene una desviacion hacia la derecha, lo cual indica que la mayoria de sus calificaciones son altas, mientras que la segunda tiene una desviacion hacia la izquierda, lo cual indica que la mayoria de sus calificaciones son bajas. La calificacion media de una es de 8 puntos de 10 y la de la otra es de 2 puntos de 10.

------------------

## c. Un gráfico tipo torta/pastel que muestre el porcentaje de comentarios de “Economy Class”, “Business Class” y “Others”, es decir, solo se deben poder observar en el gráfico 3 categorías.

In [65]:
p = df[['class']].copy().assign(class_=lambda x: np.select([x == 'Business Class', x == 'Economy Class'], ['Business', 'Economy'], default='Others'))
p = p['class_'].value_counts(normalize=True).to_frame().reset_index().rename(columns={'index': 'class', 'class_': 'percentage'})

fig = px.pie(p, values='percentage', names='class', title='Class distribution', labels={'percentage': 'Percentage', 'class': 'Class'}, color='class', color_discrete_map={'Business': 'blue', 'Economy': 'red'})
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.update_layout(width=500, height=500)
fig.show()